server/src/main/java/org/opensearch/common/blobstore/transfer/stream/OffsetRangeIndexInputStream.java

```diff
@@ -35,16 +46,50 @@ public OffsetRangeIndexInputStream(IndexInput indexInput, long size, long positi
         indexInput.seek(position);
         this.indexInput = indexInput;
         this.inputStreamIndexInput = new InputStreamIndexInput(indexInput, size);
+        ClosingStreams closingStreams = new ClosingStreams();
+        closingStreams.indexInput = indexInput;
+        closingStreams.inputStreamIndexInput = inputStreamIndexInput;
+        offsetRangeRefCount = new OffsetRangeRefCount(closingStreams);
+        closeOnce = new RunOnce(offsetRangeRefCount::decRef);
+    }
+
+    public void setClose(AtomicBoolean close) {
+        this.closed = close;
     }
 
     @Override
     public int read(byte[] b, int off, int len) throws IOException {
-        return inputStreamIndexInput.read(b, off, len);
+        ensureOpen();
+        try (OffsetRangeRefCount ignored = getStreamReference()) {
+            return inputStreamIndexInput.read(b, off, len);
+        }
+    }
+
+    private OffsetRangeRefCount getStreamReference() {
+        boolean successIncrement = offsetRangeRefCount.tryIncRef();
+        if (successIncrement == false) {
+            throw alreadyClosed("OffsetRangeIndexInputStream is already unreferenced.");
+        }
+        return offsetRangeRefCount;
+    }
+
+    private void ensureOpen() {
```

Comment:
> When close() is invoked for the first time, it will decrement the ref-count of offsetRangeRefCount and if ref count becomes 0 then closeInternal gets invoked closing the stream.

Yes that is true but `ensureOpen()` only checks the external boolean, which is never set by this class itself (unless I'm missing something).

> We need a separate atomic boolean closed because we don't want a future read to succeed when close has been invoked even if there are on-going reads

Can you solve that by keeping an internal boolean that is set when this `close()` method is invoked? Then the container class can just invoke `close()` on all its constituent streams and avoid sharing an AtomicBoolean across instances.

Reply:
Yes, one more alternative would be to keep `close` withing this class which can be set by invoking container's close method which can do so for all streams. But there the reason I prefer the original approach over this is because today we maintain all open streams in an array which is initialized on demand when a stream is created. We need the references of streams to calculate the aggregated checksum of all streams once they are fully read. Before this fix we were also using this array to close the streams iteratively. In multipart download flow, fetch from S3 is async but disk writes happen in sync. To make it async S3 exposes a file based async processor which requires a stream creator instead of the stream itself. Tomorrow, if we choose to build a custom processor or use the same file processor and lets say stream creation fails and S3 does a retry, then we will lose the reference to previous stream. And lets say due to a bug, S3 async SDK reads gets stuck or we don't receive a callback, then the stream will never get closed. 
By sharing same close flag across streams, we have a central control to close every stream that was ever created. Without this we can get into resource leak situation. With a central close flag, in the worst case (arising from S3 SDK) we may end up prematurely closing the stream which is much less severe.

Reply:
> By sharing same close flag across streams, we have a central control to close every stream that was ever created.

This is kind of my hang up here...the central flag doesn't close the stream, it sets a boolean that relies on something else to happen to trigger the operation that actually cleans up the resources. Aren't we vulnerable to a resource leak here?

Reply:
You are right! It isn't closing the stream. That came out wrong. It is setting the flag to prevent any subsequent reads on all streams. It will be the responsibility of the stream creator to close it which is what is happening now. What I wanted to say was that we can close streams iteratively in addition to the stream creator but closing streams here is just the best effort. To ensure that a stream is closed, we need to delegate the responsibility of closing the stream to place where the stream is being created.
Significance of central close would be to act as a master switch to block all subsequent reads on all created streams.

Reply:
Invoking close iteratively to set internal close flag on individual stream would just be a best effort just like closing a stream would because container doesn't have control of stream creation.

Reply:
Resolving as we decided to keep the closed flag but rename it to something like `readBlock`